### PR TITLE
Espnet engine

### PR DESCRIPTION
**Pull request by oadams**
Incorporates using ESPnet engine backend.

### CODE REVIEW

File: elpis/engines/espnet/objects/model.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.config['status'] = value
+
+    def link(self, dataset: Dataset, _pron_dict):
+        self.dataset = dataset
+        self.config['dataset_name'] = dataset.name
+        # Note the _pron_dict is ignored as it's irrelevant to ESPnet.
+
+    def build_structure(self):
+        print("BUILD STRUCTURE")
+        # NOTE Since the ESPnet data is similari informatting requirements to
```
Comment (nicklambourne) [inline elpis/engines/espnet/objects/model.py:55]: typo: similar

File: elpis/engines/espnet/objects/model.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+            src_test_dir = model_path / "output" / "testing"
+            tgt_test_dir = local_espnet_path / "data" / "test"
+            shutil.copytree(src_test_dir, tgt_test_dir)
+
+            # Then move the WAVs across
+            src_wav_dir = Path(self.dataset.path) / "resampled"
+            for wav in src_wav_dir.glob("*.wav"):
+                shutil.copy(wav, local_espnet_path)
+            #shutil.copytree(src_wav_dir, local_espnet_path, dirs_exist_ok=True)
```
Comment (nicklambourne) [inline elpis/engines/espnet/objects/model.py:101]: Does this line need to stay?

File: elpis/engines/espnet/objects/model.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+        def train():
+            local_espnet_path = Path(self.path) / 'espnet-asr1'
+            run_log_path = Path(self.path) / 'train.log'
+            print("SELF PATH {}".format(self.path))
```
Comment (nicklambourne) [inline elpis/engines/espnet/objects/model.py:106]: We tend to use fstrings rather than `.format()`.

File: elpis/engines/espnet/objects/model.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        if on_complete is None:
+            print("oncomplete is none")
+        else:
+            print("oncomplete is not none")
+            run_training_in_background()
+        return
+
+    def get_train_results(self):
+        from pathlib import Path
```
Comment (nicklambourne) [inline elpis/engines/espnet/objects/model.py:131]: This has already been imported on line 3.

File: elpis/engines/espnet/objects/model.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    def train(self, on_complete:Callable=None):
+
+        def prepare_for_training():
+            # This should just copy the ESPnet experiment directory into the
+            # model directory and then copy the prepared train/test langdir
+            # stuff into the appropriate subdirectory.
+
+            # First make a copy of the ESPNET Elpis recipe
+            model_path = Path(self.path)
+            local_espnet_path = model_path / "espnet-asr1"
```
Comment (nicklambourne) [inline elpis/engines/espnet/objects/model.py:85]: Would be great to have consistency of use between `Path() / addition` and `Path().joinpath()`, at least within this module. The rest of the project tends to use `.joinpath()`.

File: elpis/engines/espnet/objects/transcription.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        dir_util.copy_tree(f'{self.path}', f"{infer_path}")
+        #os.makedirs(f"{infer_path}", exist_ok=True)
+        #try:
+        #    shutil.rmtree(f"{infer_path}")
+        #except FileNotFoundError:
+        #    pass
+        #fs = shutil.copytree(f'{self.path}', f"{infer_path}")
+        #print("fs: {}".format(fs))
+        #print("self.path: {}".format(self.path))
+        #print("infer_path: {}".format(infer_path))
```
Comment (nicklambourne) [inline elpis/engines/espnet/objects/transcription.py:115]: Can these comments be removed?

File: elpis/engines/espnet/objects/transcription.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        # TODO Need to produce an output eaf.
+        #run(f"cd {local_espnet_path}; ./decode.sh --nj 1 --stage 0 --stop_stage 2 --recog_set infer &> {prepare_log_path}")
+        #run(f"cd {local_espnet_path}; ./decode.sh --nj 1 --stage 5 --recog_set infer &> {transcribe_log_path}")
+        # TODO Need to produce an output eaf.
+        #transcribe_json = infer_path / ".." / ".." / "exp/train_nodev_pytorch_train_mtlalpha1.0/decode_infer_decode_ctcweight1.0_nolm/data.json"
+        #file_util.copy_file(f"{transcribe_json}", f'{self.path}/one-best-hypothesis.txt')
+        #file_util.copy_file(f"{transcribe_json}", f'{self.path}/{self.hash}.eaf')
+
+        #file_util.copy_file(f"{infer_path.joinpath('utterance-0.eaf')}", f'{self.path}/{self.hash}.eaf')
+        #file_util.copy_file(f"{infer_path.joinpath('one-best-hypothesis.txt')}", f'{self.path}/one-best-hypothesis.txt')
```
Comment (nicklambourne) [inline elpis/engines/espnet/objects/transcription.py:137]: Can these be removed?

File: elpis/engines/kaldi/objects/model.py [hunk 8]
```diff
@@ -255,3 +256,31 @@ def background_train_task():
...
+                    # use line to sort by best val
+                    line_r = line.replace('%WER ', '')
+                    wer_lines.append(line_r)
+            wer_lines.sort(reverse = True)
+            line = wer_lines[0]
+            line_split = line.split(None, 1)
+            wer = line_split[0]
+            line_results = line_split[1]
+            line_results = line_results.replace('[', '')
+            line_results = line_results.replace(']', '')
```
Comment (nicklambourne) [inline elpis/engines/kaldi/objects/model.py:277]: You can use a single `re.sub()` call to replace both of these at once.

**Review by nicklambourne — CHANGES_REQUESTED** [review]
A few small things to address as well as some suggestions for better consistency with the rest of the project.

**Comment by benfoley** [comment]
Using this PR and the matching persephone-tools elpis-gui espnet-engine branch I was able to train a model successfully. 

Transcribing failed though, here's the log output:
```
172.17.0.1 - - [03/Sep/2020 03:02:45] "GET /api/transcription/transcribe HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/_compat.py", line 39, in reraise
    raise value
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/_compat.py", line 39, in reraise
    raise value
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/elpis/elpis/endpoints/transcription.py", line 34, in transcribe
    transcription.transcribe(on_complete=lambda: print('Transcribed text!'))
  File "/elpis/elpis/engines/espnet/objects/transcription.py", line 124, in transcribe
    assert len(result_paths) == 1, f"Incorrect number of result files ({len(result_path)})"
UnboundLocalError: local variable 'result_path' referenced before assignment
```

Changing `len(result_path)` to `len(result_paths)` on line 124 of `engines/espnet/objects/transcription.py` fixes this error, but `result_paths` turns out to be empty. 

This line expects a `decode_infer*` dir which isn't present after training.
```
        result_paths = list(exp_path.glob("train_nodev*/decode_infer*"))
```

File: elpis/engines/kaldi/objects/model.py (hunk 8 above)
Comment (oadams) [reply elpis/engines/kaldi/objects/model.py]: The [\[\]] in the regex is less readable but as you wish :)

**Comment by oadams** [comment]
\

> Using this PR and the matching persephone-tools elpis-gui espnet-engine branch I was able to train a model successfully.
> 
> Transcribing failed though, here's the log output:
> 
> ```
> 172.17.0.1 - - [03/Sep/2020 03:02:45] "GET /api/transcription/transcribe HTTP/1.1" 500 -
> Traceback (most recent call last):
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 2464, in __call__
>     return self.wsgi_app(environ, start_response)
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 2450, in wsgi_app
>     response = self.handle_exception(e)
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1867, in handle_exception
>     reraise(exc_type, exc_value, tb)
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/_compat.py", line 39, in reraise
>     raise value
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 2447, in wsgi_app
>     response = self.full_dispatch_request()
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1952, in full_dispatch_request
>     rv = self.handle_user_exception(e)
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1821, in handle_user_exception
>     reraise(exc_type, exc_value, tb)
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/_compat.py", line 39, in reraise
>     raise value
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1950, in full_dispatch_request
>     rv = self.dispatch_request()
>   File "/venv/lib/python3.6/site-packages/Flask-1.1.2-py3.6.egg/flask/app.py", line 1936, in dispatch_request
>     return self.view_functions[rule.endpoint](**req.view_args)
>   File "/elpis/elpis/endpoints/transcription.py", line 34, in transcribe
>     transcription.transcribe(on_complete=lambda: print('Transcribed text!'))
>   File "/elpis/elpis/engines/espnet/objects/transcription.py", line 124, in transcribe
>     assert len(result_paths) == 1, f"Incorrect number of result files ({len(result_path)})"
> UnboundLocalError: local variable 'result_path' referenced before assignment
> ```
> 
> I think there's a typo in the assertion on line 124 of `engines/espnet/objects/transcription.py`. Should `len(result_path)` be `len(result_paths)`?
> 
> But changing `len(result_path)` to `len(result_paths)` is still problematic, as `result_paths` is an empty list.
> 
> This line expects a `decode_infer*` dir which isn't present after training.
> 
> ```
>         result_paths = list(exp_path.glob("train_nodev*/decode_infer*"))
> ```
> 
> In my exp dir I have `train_nodev_pytorch_train_mtlalpha1.0` and `decode_test_decode_ctcweight1.0_nolm`.

So there's two things going on here. My code isn't crashing on the assert, which is weird. I'll look into this.

But since that isn't really your issue, I suspect the problem for your lies in the outcome of the decode.sh subprocesses. Can you share what is in `/state/models/<model-id>/prepare_transcribe_log.txt` which is the first call to decode.sh and does some data prep, and `/state/models/<model-id>/transcribe_log.txt`, which is the second and does the actual decoding?

**Review by nicklambourne — APPROVED** [review]

**Comment by nicklambourne** [comment]
**Do not merge**, code review is done but there is some fragility to be ironed out.

**Review by nicklambourne — APPROVED** [review]

**Comment by alexis-michaud** [comment]
🥇

**Pull request merged**